local-volume/provisioner/pkg/monitor/monitor.go

```diff
@@ -0,0 +1,326 @@
+		monitor.localVolumeMap.AddLocalVolume(&pv)
+	}
+	return nil
+}
+
+func (monitor *Monitor) addVolume(obj interface{}) {
+	volume, ok := obj.(*v1.PersistentVolume)
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", obj)
+		return
+	}
+
+	monitor.localVolumeMap.AddLocalVolume(volume)
+
+}
+
+func (monitor *Monitor) updateVolume(oldObj, newObj interface{}) {
+	newVolume, ok := newObj.(*v1.PersistentVolume)
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", newObj)
+		return
+	}
+
+	monitor.localVolumeMap.UpdateLocalVolume(newVolume)
+}
+
+func (monitor *Monitor) deleteVolume(obj interface{}) {
+	volume, ok := obj.(*v1.PersistentVolume)
+	if !ok {
+		glog.Errorf("Expected PersistentVolume but handler received %#v", obj)
+		return
+	}
+
+	monitor.localVolumeMap.DeleteLocalVolume(volume)
+
+}
+
+// Run starts all of this controller's control loops
+func (monitor *Monitor) Run(stopCh <-chan struct{}) {
+	glog.Infof("Starting monitor controller %s!", string(monitor.RuntimeConfig.Name))
+	monitor.hasRunLock.Lock()
+	monitor.hasRun = true
+	monitor.hasRunLock.Unlock()
+	go monitor.volumeController.Run(stopCh)
+
+	go monitor.MonitorLocalVolumes()
+	<-stopCh
+}
+
+// HasRun returns whether the volume controller has Run
+func (monitor *Monitor) HasRun() bool {
+	monitor.hasRunLock.Lock()
+	defer monitor.hasRunLock.Unlock()
+	return monitor.hasRun
+}
+
+func (monitor *Monitor) MonitorLocalVolumes() {
+	for {
+		if monitor.HasRun() {
+			pvs := monitor.localVolumeMap.GetPVs()
+			for _, pv := range pvs {
+				monitor.checkStatus(pv)
+			}
+		}
+
+		time.Sleep(DefaultMonitorResyncPeriod)
+	}
+}
+
+// checkStatus checks pv health condition
+func (monitor *Monitor) checkStatus(pv *v1.PersistentVolume) {
+	// check if PV is local storage
+	if pv.Spec.Local == nil {
+		glog.Infof("PV: %s is not local storage", pv.Name)
+		return
+	}
+	// check node and pv affinity
+	fit, err := CheckNodeAffinity(pv, monitor.Node.Labels)
+	if err != nil {
+		glog.Errorf("check node affinity error: %v", err)
+		return
+	}
+	if !fit {
+		glog.Errorf("pv: %s does not belong to this node: %s", pv.Name, monitor.Node.Name)
+		return
+	}
+
+	// check if host dir still exists
+	hostPath := pv.Spec.Local.Path
+	var rel string
+	var mountPath string
+	for _, config := range monitor.DiscoveryMap {
+		if strings.Contains(hostPath, config.HostDir) {
+			rel, err = filepath.Rel(config.HostDir, hostPath)
+			if err != nil {
+				mountPath = filepath.Join(config.MountDir, rel)
+			} else {
+				glog.Errorf("get relative path error: %v", err)
+			}
+			break
+		}
+	}
+	if len(mountPath) == 0 {
+		// can not find mount path, this may because: admin modify config(hostpath)
+		// mark PV and send a event
+		err = monitor.markPV(pv, HostPathNotExist, "yes")
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		monitor.Recorder.Event(pv, v1.EventTypeWarning, HostPathNotExist, "host path config may be modified")
```

Comment:
where is `Recorder` defined?

Reply:
at `StartLocalController` in provisioner/controller package